### PR TITLE
Consider making kcidb-validate output validated reports in kcidb-validate  #145

### LINKED ISSUE #145: Consider making kcidb-validate output validated reports
It could be useful to make `kcidb-validate` output reports after they were validated, similar to other validation tools. This could help the user see which exact report was valid and which wasn't, when it aborts on validation failure.

### PR DESCRIPTION
**Pull request by mharyam**
Close #145 

- Add print to kcidb-validate to output reports after they were validated

### CODE REVIEW

**Comment by spbnick** [comment]
Thank you, Maryam. Please "ping" me in a comment or on Slack, when you're ready for review.

**Comment by mharyam** [comment]
> Thank you, Maryam. Please "ping" me in a comment or on Slack, when you're ready for review.

Thank you 
I will

**Review by spbnick — CHANGES_REQUESTED** [review]
Yep, that's basically what needs to be done. However, could you please re-implement this using `misc.json_dump_stream()` and `misc.OutputArgumentParser`, similarly to how `upgrade_main()` does it? This way the tool would act more like others.

File: kcidb/__init__.py [hunk 1]
```diff
@@ -253,8 +253,13 @@ def validate_main():
     parser = misc.ArgumentParser(description=description)
     misc.argparse_schema_add_args(parser, "validate against")
     args = parser.parse_args()
-    for data in misc.json_load_stream_fd(sys.stdin.fileno()):
-        print(args.schema_version.validate(data))
+    misc.json_dump_stream(
+        (
+            args.schema_version.upgrade(io.schema.validate(data), copy=False)
```
Comment (spbnick) [inline kcidb/__init__.py:258]: This changes the function and the `kcidb-validate` tool to also upgrade the data, which is not in the issue you're trying to fix: https://github.com/kernelci/kcidb/issues/145

Your first attempt was basically fine, and I'm just asking you to make the tool's interface match the other tools.

Ping me on Slack, if you need help figuring this out.

File: kcidb/unittest.py [hunk 2]
```diff
@@ -85,7 +85,7 @@ def assertExecutes(self, stdin, name, *args,
         if result.returncode != status:
             errors.append(f"Expected exit status {status}, "
                           f"got {result.returncode}")
-        if not re.fullmatch(stdout_re, result.stdout, re.DOTALL):
+        if not re.match(stdout_re, result.stdout, re.DOTALL):
```
Comment (spbnick) [inline kcidb/unittest.py:88]: This is side-stepping the problem. The use of `fullmatch()` is intentional here. Something that should be documented in the function's docstring, though. Please revert this change, and if you can figure out how, document that the regular expressions are matched fully.

File: test_kcidb.py [hunk 3]
```diff
@@ -171,7 +171,8 @@ def test_validate_main(self):
         """Check kcidb-validate works"""
         self.assertExecutes('', "kcidb.validate_main")
         self.assertExecutes('{"version":{"major":4,"minor":0}}',
-                            "kcidb.validate_main")
+                            "kcidb.validate_main", "--indent=0",
+                            stdout_re='{"version": {"major": 4, "minor": 0}}')
```
Comment (spbnick) [inline test_kcidb.py:175]: Ah, you're getting there!

**Review by spbnick — CHANGES_REQUESTED** [review]
Ah, this is better :smiley: The code itself is correct, but tests need more work.

The `assertExecutes` function is quite involved. Basically, it imitates execution of a separate program given its "main" function. We need to imitate it, because we need to be able to run tests without installing KCIDB, and the executables are only created by setuptools upon installation. I.e. there are no executables in the source code. So, `assertExecutes` creates a new process calling its `main` function, feeds it the specified standard input, captures the produced stdout and stderr, and the exit status, and then checks that they match the specified requirements.

So, e.g. this line:
```
self.assertExecutes('', "kcidb.validate_main")
```
Creates a new process, which calls the `kcidb.validate_main` function in the same manner as `kcidb-validate` would, feeds it the (default) empty string through standard input (stdin), reads its standard output (stdout) and standard error (stderr) streams, as well as its exit status, and checks that they match the (default) requirements of "", "", and 0 respectively.

Hope this helps you understand what the test is doing :)

Please also try running `kcidb-validate` with various inputs both before and after your change to get a feel of what it does, and what your change does.

**Comment by spbnick** [comment]
Heey, this is very good, looks great!

However, could you please amend the commit subject to match what's being done? Something like "Output validated reports in kcidb-validate".

Also, do you think you could add another commit on top adding two more cases to the test:
* validate three valid reports in a row (in a single execution of `kcidb-validate`);
* validate three reports in a row, where the middle one is invalid, but the two others are valid?

It would be best if all three were different, in both cases, so we could make sure the ordering is correct and the right one is considered invalid.

**Comment by spbnick** [comment]
This looks very nice! Thank you, Maryam! All code and tests are fine, but could you please amend commit subjects to better describe the changes?

The first one could be something like "Output validated reports in kcidb-validate", and the second something like "tests: Add two more kcidb-validate cases".

**Comment by mharyam** [comment]
> This looks very nice! Thank you, Maryam! All code and tests are fine, but could you please amend commit subjects to better describe the changes?
> 
> The first one could be something like "Output validated reports in kcidb-validate", and the second something like "tests: Add two more kcidb-validate cases".

Done

**Comment by spbnick** [comment]
Looks good to me. Thanks a lot! Merging.

**Pull request merged**